File: packages/dart/oauth/lib/src/device_flow.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;

/// Device flow authorization for OAuth.
/// Implements the device authorization grant flow.
class DeviceFlow {
  static const String _authUrl = 'https://auth.apis.do/user_management/authorize_device';
  static const String _tokenUrl = 'https://auth.apis.do/user_management/authenticate';
  static const String defaultClientId = 'client_01JQYTRXK9ZPD8JPJTKDCRB656';

  final String clientId;
  final http.Client _client;

  /// Creates a DeviceFlow with the default client ID.
  DeviceFlow() : this.custom(defaultClientId);

  /// Creates a DeviceFlow with a custom client ID.
  DeviceFlow.custom(this.clientId) : _client = http.Client();

  /// Initiates device authorization.
  Future<DeviceAuthResponse> authorize() async {
    final response = await _client.post(
      Uri.parse(_authUrl),
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: 'client_id=$clientId',
    );

    if (response.statusCode != 200) {
      throw DeviceFlowException('Authorization failed: ${response.body}');
    }

    final json = jsonDecode(response.body) as Map<String, dynamic>;
    return DeviceAuthResponse(
      deviceCode: json['device_code'] as String,
      userCode: json['user_code'] as String,
      verificationUri: json['verification_uri'] as String,
      verificationUriComplete: json['verification_uri_complete'] as String?,
      expiresIn: json['expires_in'] as int? ?? 900,
      interval: json['interval'] as int? ?? 5,
    );
  }

  /// Polls for token after user authorization.
  Future<TokenResponse> pollForToken({
    required String deviceCode,
    int interval = 5,
    int timeout = 900,
    void Function(int)? onPoll,
  }) async {
    final startTime = DateTime.now();
    var pollInterval = interval;
    var attempts = 0;

    while (DateTime.now().difference(startTime).inSeconds < timeout) {
      attempts++;
      onPoll?.call(attempts);

      final body = 'client_id=$clientId'
          '&device_code=$deviceCode'
          '&grant_type=urn:ietf:params:oauth:grant-type:device_code';

      final response = await _client.post(
        Uri.parse(_tokenUrl),
        headers: {'Content-Type': 'application/x-www-form-urlencoded'},
        body: body,
      );

      final json = jsonDecode(response.body) as Map<String, dynamic>;

      if (json.containsKey('error')) {
        final error = json['error'] as String;
        final errorDescription = json['error_description'] as String?;

        if (error == 'slow_down') {
          pollInterval += 5;
        }

        if (error != 'authorization_pending' && error != 'slow_down') {
          return TokenResponse(
            error: error,
            errorDescription: errorDescription,
          );
        }
      } else {
        return TokenResponse(
          accessToken: json['access_token'] as String?,
          refreshToken: json['refresh_token'] as String?,
          tokenType: json['token_type'] as String? ?? 'Bearer',
          expiresIn: json['expires_in'] as int? ?? 3600,
        );
      }

      await Future.delayed(Duration(seconds: pollInterval));
    }

    throw TimeoutException('Authorization timed out after $timeout seconds');
  }

  /// Closes the HTTP client.
  void close() {
    _client.close();
  }

  /// Gets the client ID.
  String getClientId() => clientId;
}

/// Device authorization response.
class DeviceAuthResponse {
  final String deviceCode;
  final String userCode;
  final String verificationUri;
  final String? verificationUriComplete;
  final int expiresIn;
  final int interval;

  DeviceAuthResponse({
    required this.deviceCode,
    required this.userCode,
    required this.verificationUri,
    this.verificationUriComplete,
    this.expiresIn = 900,
    this.interval = 5,
  });

  @override
  String toString() => 'Visit $verificationUri and enter code: $userCode';
}

/// Token response from authentication.
class TokenResponse {
  final String? accessToken;
  final String? refreshToken;
  final String tokenType;
  final int expiresIn;
  final String? error;
  final String? errorDescription;

  TokenResponse({
    this.accessToken,
    this.refreshToken,
    this.tokenType = 'Bearer',
    this.expiresIn = 3600,
    this.error,
    this.errorDescription,
  });

  bool get isError => error != null;
  bool get isPending => error == 'authorization_pending' || error == 'slow_down';
}

/// Device flow exception.
class DeviceFlowException implements Exception {
  final String message;
  DeviceFlowException(this.message);

  @override
  String toString() => message;
}

File: packages/php/oauth/src/DeviceFlow.php
<?php

declare(strict_types=1);

namespace DotDo\OAuth;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;

/**
 * Device flow authentication.
 */
class DeviceFlow
{
    public const DEFAULT_CLIENT_ID = 'client_01JQYTRXK9ZPD8JPJTKDCRB656';

    private const AUTH_URL = 'https://auth.apis.do/user_management/authorize_device';
    private const TOKEN_URL = 'https://auth.apis.do/user_management/authenticate';
    private const USER_URL = 'https://apis.do/me';

    private Client $client;
    private string $clientId;

    public function __construct(string $clientId = self::DEFAULT_CLIENT_ID)
    {
        $this->clientId = $clientId;
        $this->client = new Client([
            'timeout' => 30,
            'http_errors' => false,
        ]);
    }

    /**
     * Initiates device authorization.
     *
     * @param string $scope OAuth scope
     * @return array Device authorization response
     * @throws \Exception on failure
     */
    public function authorize(string $scope = 'openid profile email'): array
    {
        $response = $this->client->post(self::AUTH_URL, [
            'json' => [
                'client_id' => $this->clientId,
                'scope' => $scope,
            ],
        ]);

        $statusCode = $response->getStatusCode();
        $body = json_decode((string) $response->getBody(), true);

        if ($statusCode !== 200) {
            throw new \Exception("Authorization failed: " . ($body['error'] ?? 'Unknown error'));
        }

        return $body;
    }

    /**
     * Polls for tokens until user authorizes or timeout.
     *
     * @param string $deviceCode Device code
     * @param int $interval Polling interval in seconds
     * @param int $expiresIn Expiration time in seconds
     * @return array Token response
     * @throws \Exception on failure
     */
    public function pollForTokens(string $deviceCode, int $interval, int $expiresIn): array
    {
        $deadline = time() + $expiresIn;

        while (time() < $deadline) {
            $response = $this->client->post(self::TOKEN_URL, [
                'json' => [
                    'client_id' => $this->clientId,
                    'device_code' => $deviceCode,
                    'grant_type' => 'urn:ietf:params:oauth:grant-type:device_code',
                ],
            ]);

            $statusCode = $response->getStatusCode();
            $body = json_decode((string) $response->getBody(), true);

            if ($statusCode === 200) {
                return $body;
            }

            if ($statusCode === 400) {
                $error = $body['error'] ?? '';

                if ($error === 'authorization_pending') {
                    sleep($interval);
                    continue;
                }

                if ($error === 'slow_down') {
                    $interval += 5;
                    sleep($interval);
                    continue;
                }

                throw new \Exception("Token request failed: $error");
            }

            throw new \Exception("Unexpected response: $statusCode");
        }

        throw new \Exception('Authorization expired');
    }

    /**
     * Gets user info with access token.
     *
     * @param string $accessToken Access token
     * @return array User info
     * @throws \Exception on failure
     */
    public function getUserInfo(string $accessToken): array
    {
        $response = $this->client->get(self::USER_URL, [
            'headers' => [
                'Authorization' => "Bearer $accessToken",
            ],
        ]);

        $statusCode = $response->getStatusCode();
        $body = json_decode((string) $response->getBody(), true);

        if ($statusCode !== 200) {
            throw new \Exception("Failed to get user info: " . ($body['error'] ?? 'Unknown error'));
        }

        return $body;
    }
}
